ARCs/arc-0022.md

```diff
@@ -0,0 +1,64 @@
+---
+arc: 22
+title: Extend arc-4 to add `read-only` annotation to methods
+status: Draft
+---
+
+# Extend arc-4 to add `read-only` annotation to methods
+
+## Summary
+
+The following document introduces a convention for creating methods (as described in [ARC-4](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0004.md)) which don't mutate state.
+
+## Abstract
+
+The goal of this convention is to allow smart contract developers to distinguish between methods which mutate state and methods which don't by introducing a new property to the `Method` descriptor.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+> Comments like this are non-normative.
+
+## Read-only functions
+
+A `read-only` function is a function with no side-effects. In particular, a `read-only` function **SHOULD NOT** include:
+- local/global state modifications
+- calls to non `read-only` functions
+- inner-transactions
```

Comment:
Does this mean a read-only function cannot call another read-only function from another contract? I can see arguments for and against this, but on the whole I think it would be a lot more powerful if read-only functions could do this

Reply:
I agree we should allow that. Maybe, we can make it clearer this way:

Here is a proposal: replace 

```
- calls to non `read-only` functions
- inner-transactions
```

with

```
- inner transactions, except inner transactions that are application calls to `read only` methods of this smart contract or another smart contract
```

Reply:
What is the use case for either pure or readonly? Why would developers or IDE developers need have this expressed as a specific boolean instead of the information being available in the description? 

My thoughts are if there's a real need for this then it would be made more meaningful if the app call transaction readonly effect was enforced by the protocol (eg. write to global would fail).  Then in that case you might also want ambiguous methods, where write side effects are driven by arguments, then you might want to be able to call the method in 'read only mode' by specifiying that in a new property of the app call txn

Reply:
The reason for read-only would be allow DApps to query a smart contract without having to pay transaction fees.

Reply:
How does the read only flag help with that?